conda_libmamba_solver/shards_subset.py

```diff
@@ -257,3 +407,160 @@ def build_repodata_subset(
+        node_ids = batch[:]
+        with suppress(queue.Empty):
+            while True:  # loop exits with break or queue.Empty exception
+                batch = in_queue.get_nowait()
+                if batch is None:
+                    # do the work but then quit
+                    running = False
+                    break
+                else:
+                    node_ids.extend(batch)
+        yield node_ids
+
+
+def cache_fetch_thread(
+    in_queue: Queue[list[NodeId] | None],
+    shard_out_queue: Queue[list[tuple[NodeId, ShardDict]] | None],
+    network_out_queue: Queue[list[NodeId] | None],
+    cache: ShardCache,
+):
+    """
+    Fetch batches of shards from cache until in_queue sees None. Enqueue found
+    shards to shard_out_queue, and not found shards to network_out_queue.
+
+    When we see None on in_queue, we send None to both out queues and exit.
+    """
+    cache = cache.copy()
+
+    for node_ids in combine_batches_until_none(in_queue):
+        cached = cache.retrieve_multiple([node_id.shard_url for node_id in node_ids])
+
+        # should we add this into retrieve_multiple?
+        found: list[tuple[NodeId, ShardDict]] = []
+        not_found: list[NodeId] = []
+        for node_id in node_ids:
+            if shard := cached.get(node_id.shard_url):
+                found.append((node_id, shard))
+            else:
+                not_found.append(node_id)
+
+        # Might wake up the network thread by calling it first:
+        if not_found:
+            network_out_queue.put(not_found)
+        if found:
+            shard_out_queue.put(found)
+
+    network_out_queue.put(None)
+    shard_out_queue.put(None)
+
+
+def network_fetch_thread(
+    in_queue: Queue[list[NodeId] | None],
+    shard_out_queue: Queue[list[tuple[NodeId, ShardDict]] | None],
+    # how to communicate errors?
+    cache: ShardCache,
+    shardlikes: list[ShardBase],
+):
+    """
+    in_queue contains urls to fetch over the network.
+    While the in_queue has not received a sentinel None, empty everything from
+    the queue. Fetch all of them over the network. Fetched shards go to
+    shard_out_queue.
+    """
+    cache = cache.copy()
+    dctx = zstandard.ZstdDecompressor(max_window_size=ZSTD_MAX_SHARD_SIZE)
+
+    shardlikes_by_url = {s.url: s for s in shardlikes}
+
+    def fetch(s, url: str, node_id: NodeId):
+        response = s.get(url)
+        response.raise_for_status()
+        data = response.content
+        return (url, node_id, data)
+
+    def submit(node_id):
+        # this worker should only recieve network node_id's:
+        shardlike = shardlikes_by_url[node_id.channel]
+        if not isinstance(shardlike, Shards):
+            log.warning("network_fetch_thread got non-network shardlike")
+            return
+        session = shardlike.session
+        url = shardlikes_by_url[node_id.channel].shard_url(node_id.package)
+        return executor.submit(fetch, session, url, node_id)
```

Comment:
This relies on the fact that the with statement doesn't create a scope. Will probably change it, but thought that was interesting.